### ISSUE TITLE
After SFT warm-up, Model only generate <short>?

### ISSUE THREAD

**Issue by 541wsy** [VainF/Thinkless#5]
Hello, thank you for your excellent work. However, I have some questions about the warm-up. According to the settings in your paper, I used our own business data with approximately 7k prompts, generating one answer each for think and short modes. The base model was deepseek-distill-qwen-1.5b, but after SFT, I found that the model only outputs responses in short mode. Have you encountered a similar situation?

By the way, did you add a <short> token to the tokenizer?

**Comment by VainF** [comment]
Hi @541wsy, Have you checked the chat template? Maybe there is a <short> in the chat template. 

> By the way, did you add a token to the tokenizer?

We did not add a new token. Instead, we take a predefined but unused token as the <short>.

**Comment by 541wsy** [comment]
Thanks for your reply!
Could you share your chat_template? I noticed that the default chat template for deepseek-distill-qwen-1.5b automatically adds the <think> token—should this token be removed?

Also, could you provide a concrete example to explain what "predefined but unused token" refers to? Thank you very much!?

**Comment by 541wsy** [comment]
> Hi [@541wsy](https://github.com/541wsy), Have you checked the chat template? Maybe there is a `<short>` in the chat template.
> 
> > By the way, did you add a token to the tokenizer?
> 
> We did not add a new token. Instead, we take a predefined but unused token as the `<short>`.

Thanks for your reply!
Could you share your chat_template? I noticed that the default chat template for deepseek-distill-qwen-1.5b automatically adds the token—should this token be removed?

Also, could you provide a concrete example to explain what "predefined but unused token" refers to? Thank you very much!?

**Comment by Rinnolo** [comment]
I was confused about it too.I have read the paper and I notice that the authors mention that "they found that the RL will also compress the length of long responses, since the algorithm will encourage short and correct answers, producing a gradient towards a more compact response." . But in Reward setting there has not any reward about length, only reward  about the control token. 

So if possible, there may be a reward hack that model always generate "<short>" token followed by a long CoT to generate a correct answer? Or the SFT phase can make model know that if it generate "<short>" token and then it will always generate a short CoT ,which I think impracticable perhaps. 

Hope for replying. ovo

**Comment by VainF** [comment]
> > Hi [@541wsy](https://github.com/541wsy), Have you checked the chat template? Maybe there is a `<short>` in the chat template.
> > > By the way, did you add a token to the tokenizer?
> > 
> > 
> > We did not add a new token. Instead, we take a predefined but unused token as the `<short>`.
> 
> Thanks for your reply! Did you modify the chat template of deepseek-distill-qwen-1.5b?
> 
> Also, could you provide a concrete example to explain what "predefined but unused token" refers to? Thank you very much!

Here is an example: 

https://huggingface.co/Vinnnf/Thinkless-1.5B-Warmup/blob/main/tokenizer_config.json#L62

We replaced the token 151650 with the `<short>`. In addition, we also remove the hard-coded `<think>` at the end of the chat template:

https://huggingface.co/Vinnnf/Thinkless-1.5B-Warmup/blob/main/tokenizer_config.json#L184

**Comment by CurryxIaoHu** [comment]
> Hello, thank you for your excellent work. However, I have some questions about the warm-up. According to the settings in your paper, I used our own business data with approximately 7k prompts, generating one answer each for think and short modes. The base model was deepseek-distill-qwen-1.5b, but after SFT, I found that the model only outputs responses in short mode. Have you encountered a similar situation?
> 
> By the way, did you add a token to the tokenizer?

The problem I have is that the model will only output <think>...

**Comment by 541wsy** [comment]
> > Hello, thank you for your excellent work. However, I have some questions about the warm-up. According to the settings in your paper, I used our own business data with approximately 7k prompts, generating one answer each for think and short modes. The base model was deepseek-distill-qwen-1.5b, but after SFT, I found that the model only outputs responses in short mode. Have you encountered a similar situation?
> > By the way, did you add a token to the tokenizer?
> 
> The problem I have is that the model will only output `<think>`...

I also encountered that problem with a small lr (1e-6). When increasing the lr to 5e-6, the model only output short. BTW, I finally solve the problem by upsampling the think sample in the sft trainset.

**Comment by VainF** [comment]
The Megatron-LM recipe used for SFT. Hope this will be helpful. 

By the way, we ensure that all questions have both short and long responses in each batch. 

```bash
#!/usr/bin/bash

# Get Data Blend
DATA_BLEND="assets/data/OpenThoughts-1M/Tokenized-open_thoughts_1m_math_hybrid-deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B-16384_text_document"
export MASTER_ADDR="127.0.0.1" # select the master address
export MASTER_PORT="45521" # select the port

# Device Configs
NNODES=1 # number of nodes. 
NPROC_PER_NODE=4 # number of gpus (processes) per node
export WORLD_SIZE=$(($NNODES * $NPROC_PER_NODE)) 
mode=$1 # resume from checkpoint

# Task Configs
TAG="Hybrid-R1-1.5B-OpenThoughts-1M-Math-1.5B-Distill" # this will be the name of output folder
DATA_INDEX_PATH=CACHE # path to the cache folder. Will generate if not exists
PROJECT_PATH=$(pwd)
OUTPUT_PATH="$PROJECT_PATH/outputs"

# Transformer Configs
HIDEN_SIZE=1536 # hidden size
NUM_LAYERS=28 # number of layers
NUM_ATTN_HEADS=12 # number of attention heads
NUM_QUERY_GROUPS=2 # number of query groups
FFN_HIDDEN_SIZE=8960 # feed forward network hidden size
RMS_NORM_EPS=1e-6 # epsilon for rms layer normalization
SEQ_LENGTH=16384 # sequence length
MAX_POSITION_EMBEDDINGS=32768 # maximum position embeddings
ROPE_BASE=10000

# Training Configs
TOKENIZER_MODEL="$PROJECT_PATH/assets/checkpoints/deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B" # path to the tokenizer model

TENSOR_PARALLEL_SIZE=1
PIPELINE_PARALLEL_SIZE=1
CONTEXT_PARALLEL_SIZE=1
LR=1e-5
MIN_LR=1e-6
TRAIN_ITERS=17000 # number of iterations to train for
WARMUP_ITERS=$(expr $TRAIN_ITERS \* 5 / 100)
MICRO_BATCH_SIZE=1
GLOBAL_BATCH_SIZE=128

# intervals
SAVE_INTERVALS=1000
LOG_INTERVALS=10
EVAL_INTERVALS=100
EVAL_ITERS=10

# Set Training configs
CKPT_SUBDIR="$OUTPUT_PATH/checkpoints/$TAG/train_iters_$TRAIN_ITERS"
if [ "$mode" = "train" ]; then
    LOAD="assets/checkpoints/deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B_Megatron_TP${TENSOR_PARALLEL_SIZE}PP${PIPELINE_PARALLEL_SIZE}"
    EXTRA_CMD="--no-load-optim --no-load-rng --finetune --load $LOAD" 
else
    LOAD="$CKPT_SUBDIR"
    EXTRA_CMD="--load $LOAD"
fi  

cd $PROJECT_PATH; mkdir -p $CKPT_SUBDIR; export WANDB_API_KEY=$WANDB_API_KEY; echo Start Training

# cp current script to the output folder
cp $0 $CKPT_SUBDIR

OPTIONS=" \
--wandb-project Tiny-O1 \
--wandb-exp-name $TAG \
--dataloader-type cyclic \
--add-qkv-bias \
--disable-bias-linear \
--swiglu \
--attention-dropout 0.0 \
--hidden-dropout 0.0 \
--position-embedding-type rope \
--rotary-base $ROPE_BASE \
--tensor-model-parallel-size $TENSOR_PARALLEL_SIZE \
--pipeline-model-parallel-size $PIPELINE_PARALLEL_SIZE \
--context-parallel-size $CONTEXT_PARALLEL_SIZE \
--num-layers $NUM_LAYERS  \
--hidden-size $HIDEN_SIZE \
--num-attention-heads $NUM_ATTN_HEADS \
--seq-length $SEQ_LENGTH \
--max-position-embeddings $MAX_POSITION_EMBEDDINGS \
--ffn-hidden-size $FFN_HIDDEN_SIZE --normalization RMSNorm \
--micro-batch-size $MICRO_BATCH_SIZE \
--global-batch-size $GLOBAL_BATCH_SIZE \
--group-query-attention \
--num-query-groups $NUM_QUERY_GROUPS \
--train-iters $TRAIN_ITERS \
--lr $LR \
--min-lr $MIN_LR \
--lr-decay-style cosine \
--log-interval $LOG_INTERVALS \
--eval-iters $EVAL_ITERS \
--eval-interval $EVAL_INTERVALS \
--data-path "$DATA_BLEND"  \
--data-cache-path $DATA_INDEX_PATH \
--tokenizer-type HuggingFaceTokenizer \
--tokenizer-model ${TOKENIZER_MODEL} \
--save-interval $SAVE_INTERVALS \
--save $CKPT_SUBDIR \
--split 999,1,0 \
--clip-grad 1.0 \
--weight-decay 0.1 \
--adam-beta1 0.9 \
--adam-beta2 0.95 \
--init-method-std 0.02  \
--log-num-zeros-in-grad \
--lr-warmup-iters $WARMUP_ITERS \
--exit-on-missing-checkpoint \
--use-flash-attn \
--bf16 \
--exit-signal-handler \
--ckpt-format torch \
--rotary-percent 1.0 \
--calculate-per-token-loss \
--rotary-seq-len-interpolation-factor 1 \
--norm-epsilon $RMS_NORM_EPS \
--use-distributed-optimizer \
--overlap-param-gather \
--overlap-grad-reduce \
--sft \
--untie-embeddings-and-output-weights \
${EXTRA_CMD}"

export CUDA_DEVICE_MAX_CONNECTIONS=1;

torchrun --nproc_per_node=$NPROC_PER_NODE --nnodes=$NNODES --master_addr=$MASTER_ADDR --master_port=$MASTER_PORT pretrain_gpt.py ${OPTIONS}
```

**Comment by 541wsy** [comment]
> The Megatron-LM recipe used for SFT. Hope this will be helpful.
> 
> By the way, we ensure that all questions have both short and long responses in each batch. Not sure if this is important for balanced learning.
> 
> #!/usr/bin/bash
> 
> # Get Data Blend
> DATA_BLEND="assets/data/OpenThoughts-1M/Tokenized-open_thoughts_1m_math_hybrid-deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B-16384_text_document"
> export MASTER_ADDR="127.0.0.1" # select the master address
> export MASTER_PORT="45521" # select the port
> 
> # Device Configs
> NNODES=1 # number of nodes. 
> NPROC_PER_NODE=4 # number of gpus (processes) per node
> export WORLD_SIZE=$(($NNODES * $NPROC_PER_NODE)) 
> mode=$1 # resume from checkpoint
> 
> # Task Configs
> TAG="Hybrid-R1-1.5B-OpenThoughts-1M-Math-1.5B-Distill" # this will be the name of output folder
> DATA_INDEX_PATH=CACHE # path to the cache folder. Will generate if not exists
> PROJECT_PATH=$(pwd)
> OUTPUT_PATH="$PROJECT_PATH/outputs"
> 
> # Transformer Configs
> HIDEN_SIZE=1536 # hidden size
> NUM_LAYERS=28 # number of layers
> NUM_ATTN_HEADS=12 # number of attention heads
> NUM_QUERY_GROUPS=2 # number of query groups
> FFN_HIDDEN_SIZE=8960 # feed forward network hidden size
> RMS_NORM_EPS=1e-6 # epsilon for rms layer normalization
> SEQ_LENGTH=16384 # sequence length
> MAX_POSITION_EMBEDDINGS=32768 # maximum position embeddings
> ROPE_BASE=10000
> 
> # Training Configs
> TOKENIZER_MODEL="$PROJECT_PATH/assets/checkpoints/deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B" # path to the tokenizer model
> 
> TENSOR_PARALLEL_SIZE=1
> PIPELINE_PARALLEL_SIZE=1
> CONTEXT_PARALLEL_SIZE=1
> LR=1e-5
> MIN_LR=1e-6
> TRAIN_ITERS=17000 # number of iterations to train for
> WARMUP_ITERS=$(expr $TRAIN_ITERS \* 5 / 100)
> MICRO_BATCH_SIZE=1
> GLOBAL_BATCH_SIZE=128
> 
> # intervals
> SAVE_INTERVALS=1000
> LOG_INTERVALS=10
> EVAL_INTERVALS=100
> EVAL_ITERS=10
> 
> # Set Training configs
> CKPT_SUBDIR="$OUTPUT_PATH/checkpoints/$TAG/train_iters_$TRAIN_ITERS"
> if [ "$mode" = "train" ]; then
>     LOAD="assets/checkpoints/deepseek_ai_DeepSeek_R1_Distill_Qwen_1.5B_Megatron_TP${TENSOR_PARALLEL_SIZE}PP${PIPELINE_PARALLEL_SIZE}"
>     EXTRA_CMD="--no-load-optim --no-load-rng --finetune --load $LOAD" 
> else
>     LOAD="$CKPT_SUBDIR"
>     EXTRA_CMD="--load $LOAD"
> fi  
> 
> cd $PROJECT_PATH; mkdir -p $CKPT_SUBDIR; export WANDB_API_KEY=$WANDB_API_KEY; echo Start Training
> 
> # cp current script to the output folder
> cp $0 $CKPT_SUBDIR
> 
> OPTIONS=" \
> --wandb-project Tiny-O1 \
> --wandb-exp-name $TAG \
> --dataloader-type cyclic \
> --add-qkv-bias \
> --disable-bias-linear \
> --swiglu \
> --attention-dropout 0.0 \
> --hidden-dropout 0.0 \
> --position-embedding-type rope \
> --rotary-base $ROPE_BASE \
> --tensor-model-parallel-size $TENSOR_PARALLEL_SIZE \
> --pipeline-model-parallel-size $PIPELINE_PARALLEL_SIZE \
> --context-parallel-size $CONTEXT_PARALLEL_SIZE \
> --num-layers $NUM_LAYERS  \
> --hidden-size $HIDEN_SIZE \
> --num-attention-heads $NUM_ATTN_HEADS \
> --seq-length $SEQ_LENGTH \
> --max-position-embeddings $MAX_POSITION_EMBEDDINGS \
> --ffn-hidden-size $FFN_HIDDEN_SIZE --normalization RMSNorm \
> --micro-batch-size $MICRO_BATCH_SIZE \
> --global-batch-size $GLOBAL_BATCH_SIZE \
> --group-query-attention \
> --num-query-groups $NUM_QUERY_GROUPS \
> --train-iters $TRAIN_ITERS \
> --lr $LR \
> --min-lr $MIN_LR \
> --lr-decay-style cosine \
> --log-interval $LOG_INTERVALS \
> --eval-iters $EVAL_ITERS \
> --eval-interval $EVAL_INTERVALS \
> --data-path "$DATA_BLEND"  \
> --data-cache-path $DATA_INDEX_PATH \
> --tokenizer-type HuggingFaceTokenizer \
> --tokenizer-model ${TOKENIZER_MODEL} \
> --save-interval $SAVE_INTERVALS \
> --save $CKPT_SUBDIR \
> --split 999,1,0 \
> --clip-grad 1.0 \
> --weight-decay 0.1 \
> --adam-beta1 0.9 \
> --adam-beta2 0.95 \
> --init-method-std 0.02  \
> --log-num-zeros-in-grad \
> --lr-warmup-iters $WARMUP_ITERS \
> --exit-on-missing-checkpoint \
> --use-flash-attn \
> --bf16 \
> --exit-signal-handler \
> --ckpt-format torch \
> --rotary-percent 1.0 \
> --calculate-per-token-loss \
> --rotary-seq-len-interpolation-factor 1 \
> --norm-epsilon $RMS_NORM_EPS \
> --use-distributed-optimizer \
> --overlap-param-gather \
> --overlap-grad-reduce \
> --sft \
> --untie-embeddings-and-output-weights \
> ${EXTRA_CMD}"
> 
> export CUDA_DEVICE_MAX_CONNECTIONS=1;
> 
> torchrun --nproc_per_node=$NPROC_PER_NODE --nnodes=$NNODES --master_addr=$MASTER_ADDR --master_port=$MASTER_PORT pretrain_gpt.py ${OPTIONS}

thanks for your help!
